[ros2interface] Allow stdin input for 'ros2 interface show'

sharminramli opened the pull request:
Implements #349.

Interface types can now be piped directly into show, like this:
`$ ros2 topic type /chatter | ros2 interface show`

sharminramli commented:
I've attached an action to `add_argument` that reads from stdin if `-` was passed. Requiring `-` preserves the parser's `required` argument check on `type` since we don't have to allow `type` to be missing. This requires explicitly running

`$ ros2 topic type /chatter | ros2 interface show -`

rather than

`$ ros2 topic type /chatter | ros2 interface show`

This also replaces reading `stdin` directly for the default value and fixes the resulting exceptions.

I've also added more information in the help but I'm not sure if it's written clearly enough, or if there's a convention that I'm breaking.

dirk-thomas commented:
Thanks for the updates.

I applies a few changes. Please have a look:
* 0134dd07ccb5d0c5486b7e1c4e1a58715371ffe6 Move all the help text into the `help` attribute. That avoids potentially overwriting the epilog if set externally. Also the `textwrap.dedent` seems unnecessary.
* a085e270b6ec5f9e8e031e0844a6ba39f46e355b Fix the import order to pass the linters.
* f14c359c4456e2b141fbfb9eaad55dbcf2608dd7 Simplify the logic in the custom action to avoid duplicate parts.
* 3e7241a0818c79fe5f8795980b4092cb7acadfea Handle the case where the piped command fails / doesn't output anything more gracefully.

dirk-thomas commented:
It would be great if you could add a unit test to cover this new feature to avoid it regressing in the future.

dirk-thomas commented:
> It would be great if you could add a unit test to cover this new feature to avoid it regressing in the future.

@sharminramli Friendly ping.

sharminramli commented:
@dirk-thomas Sorry about the delay. Thanks for the changes, they look great.

I started working on the unit tests, but I ran into some issues piping stdin to the test process. I am patching a mock stdin to the test function with
```
    @patch('sys.stdin', new_callable=StringIO)
    def test_show_message_from_stdin(self, mocked_stdin):
        mocked_stdin.write('std_msgs/msg/String')
        mocked_stdin.seek(0)
        with self.launch_interface_command(
            arguments=['show', '-']
        ) as interface_command:
            assert interface_command.wait_for_shutdown(timeout=2)
        assert interface_command.exit_code == launch_testing.asserts.EXIT_OK
        assert launch_testing.tools.expect_output(
            expected_lines=[
                'string data'
            ],
            text=interface_command.output,
            strict=True
        )
```
but it appears to hang at
`values = sys.stdin.readline().strip()` of `show.py`. Is there something I'm missing?

hidmic commented:
@sharminramli you won't be able to pipe data into the launched process like that. `launch` does not expose the `stdin` pipe either, so your best bet right now is to use `shell=True` and pipe. For that, you'll have to change the test fixture a bit to be able to launch an action like:

```python
command_action = ExecuteProcess(
  cmd=[sys.executable, '-c', "print('std_msgs/msg/String')", '|', 'ros2', 'interface', 'show', '-'],
  shell=True,
  output='screen'
)
``` 

Alternatively, you can propose a reasonable to expose the `stdin` pipe for launched processes in general.

dirk-thomas commented:
@sharminramli Friendly ping.

sharminramli commented:
@dirk-thomas Sorry for the wait. I've added the test as `test_show_stdin` in `test_cli.py`.
@hidmic Thanks for the help. I'm not familiar with the `launch` repo as of now, so I've gone with the first option and modified `launch_interface_command` to accept the `shell` option and a list of arguments to be prepended.

sharminramli commented:
Rebased and changed the test to use `test_msgs/msg/Nested` rather than `std_msgs/msg/String` (https://github.com/ros2/common_interfaces/issues/89). The help message should also be updated but it's being modified in https://github.com/ros2/ros2cli/pull/516 as well.

hidmic commented:
> however, `test_show_message` fails because of the comments added to the message file in Foxy.

That was fixed in #516. As a result, this now needs a rebase.

sharminramli commented:
Rebased and changed the help message to refer to `example_interfaces` rather than `std_msgs`. I'm not sure why the Jenkins build is failing though.

audrow commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10895)](http://ci.ros2.org/job/ci_linux/10895/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6251)](http://ci.ros2.org/job/ci_linux-aarch64/6251/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8873)](http://ci.ros2.org/job/ci_osx/8873/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10787)](http://ci.ros2.org/job/ci_windows/10787/)

hidmic commented:
Thanks for re-running CI @audrow !

hidmic commented:
Alright, thank you for contribution and your patience @sharminramli ! Merging.
